### DISCUSSION TITLE
how to implement role based permissions

### DISCUSSION THREAD

**Discussion by RepComm** [pocketbase/pocketbase#1850] [Q&A] [answered]
Hi, I'm trying to create a database with a specific structure:

- categories
  - boards
    - threads
      - replies
- users
  - roles
    - permissions [name, description, type (allow|disallow)]

I would like to then create some entries for API rules that can filter list/view/create/delete of
- threads
- replies

based on the user's roles permissions

Preferably I'd write something like

threads.list -> @request.auth.roles?.permissions ?~ permissions[ thread.board.name + "visible" ]
or
Allow listing threads of a specific board if at least one of the requester's roles contains a permission with the same name as (the thread's board name + "_visible")

Is something like this possible with just API rules?
Another option I thought of was to have an admin bot listen to updates and undo actions that didn't match this behavior. Possibly with a boolean that shows 'admin bot approved', which clients can subscribe to only approved updates.

Any thoughts?

**Comment by RepComm** [comment]
I am a big dumb and didn't notice the discussions feature of github or that it was in heavy use for this repo. Sorry lads.

**Comment by ganigeorgiev** [comment, marked as answer]
Concatenations and other dynamic expressions are currently not supported in the API rules syntax (_mostly for optimizations_). More about the API rules and filter syntax you can find in https://pocketbase.io/docs/api-rules-and-filters/#filters-syntax.

I'm not sure that I understand your use case and why there is a need to always concatenate `"_visible"`. If this is supposed to serve as a flag, then you can just use a `bool` field to indicate the visibility state.

There are several type of relations structure you can design with PocketBase (single, multiple, via/junction collection, etc.) and it is difficult to recommend something without the specifics of your use cases and the relations between the collections.


If I'm reading correctly your example, you have a `users` collection with single-`relation` `roles` field (eg. `role`) pointing to the `roles` collection that have another multi-`relation` `permisssions` field  pointing to the `permissions` collection, right?

I'm also assuming that the `threads` collection has a single-`relation` field, eg. `board`, to the `boards` collection.

In that case if you want to check the authenticated user permissions from the `threads` collections, then you can use something like:

```js
@request.auth.id != "" &&
@request.auth.role.permissions.name ?= board.name &&
@request.auth.role.permissions.type ?= "allow"
```

The above rule roughly translates to:

- require the request client to be authenticated
- require the authenticated client to have at least one role with permission's name matching the thread's `board.name`
- require the matched permission to also have `type=allow`

In general, you can think about the dot-notation syntax as a way to access each element from the relation(s) and its fields.
The `?` prefix infront of the equal operator means "any/at least one of" match.
If the operator was only `=`, then this will require _every related_ permission within _every related_ role to match the other operand.

**Reply by TobiasNickel** [reply]
could something like this be documented on the website? I also had to do lots of experiments: making sure that for the `permissions` type check, the same row as for the name check is used. not that he has some permission with that name, but an other permission with `type. allow`. 

This stuff is awesome and more flexible than the community edition of strapi.

**Comment by RepComm** [comment]
Thanks for the fast reply!

The way I have it set up now is with permission being named for what they suggest to do in the API rules: can_reply, can_delete, can_delete_own, can_create_thread, etc

As well as roles being plural for every user (I assumed this wouldn't work, but seeing what you did with permissions now I'm thinking it will)

At this point I'm thinking the roles (which I hope to make dynamically added like discord) should reference the boards they apply to, that way the permissions can just be checked with = "can_reply"

So:
@request.auth.id != "" &&
@request.auth.roles.board.id ?= board.id &&
@request.auth.roles.permissions.name ?= "can_reply" &&
@request.auth.role.permissions.type ?= "allow"

Does this make sense?
I'll be trying out some things tomorrow. Again thank you very much for your help!

**Reply by magikmw** [reply]
Hi, how did it go? I'm looking into trying pocketbase and RBAC is one of my usecase's required features. Would love any leads.

**Reply by cheskoxd** [reply]
What about using permissions as an options, and then checking that at least one of the permissions meets can_reply 

Idk just thought about that
